### PR TITLE
Dropping vanished followers works again.

### PR DESCRIPTION
**Pull request by neunhoef**
An exception response to a replication request is now handled properly, such that a vanished follower can properly be dropped and the leader can move on. This functionality is covered by the node resilience tests, which lead to the discovery of this bug.

### CODE REVIEW

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11451/

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11452/

File: arangod/Network/Methods.cpp [hunk 1]
```diff
@@ -234,6 +234,13 @@ class RequestsState final : public std::enable_shared_from_this<RequestsState> {
...
   void startRequest() {
+    if (!_pool) {
+      LOG_TOPIC("5949f", ERR, Logger::COMMUNICATION)
+          << "connection pool unavailable";
+      callResponse(Error::Canceled, nullptr, std::move(_request));
+      return;
+    }
+
     auto now = std::chrono::steady_clock::now();
     if (now > _endTime || _pool->config().clusterInfo->server().isStopping()) {
```
Comment (mchacki) [inline arangod/Network/Methods.cpp:245]: very good, fixed this hidden nullptr issue.
Although I do not think it actually happens.

File: arangod/Network/Methods.cpp [hunk 2]
```diff
@@ -234,6 +234,13 @@ class RequestsState final : public std::enable_shared_from_this<RequestsState> {
 
   // scheduler requests that are due
   void startRequest() {
+    if (!_pool) {
```
Comment (mchacki) [inline arangod/Network/Methods.cpp:237]:
```suggestion
    if (ADB_UNLIKELY(!_pool)) {
```

I think in reality, we will never loose the pool.

Comment (mchacki) [reply arangod/Network/Methods.cpp:237]: But feel free to not include this, the code you have is correct, this is just some very minor perf optimization which will not be measurable. (as opposed to the network we are going to use)

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11453/

File: arangod/Transaction/Methods.cpp [hunk 3]
```diff
@@ -2343,19 +2343,27 @@ Future<Result> Methods::replicateOperations(
...
-                            resp.response->statusCode() == fuerte::StatusCreated ||
-                            resp.response->statusCode() == fuerte::StatusOK;
-        if (replicationWorked) {
-          bool found;
-          resp.response->header.metaByKey(StaticStrings::ErrorCodes, found);
-          replicationWorked = !found;
+      try {
+        // Note that it is possible that the future was resolved with an exception
+        // which will be thrown when we ask for the response, we need to catch
+        // the exception here.
```
Comment (danielhlarkin) [inline arangod/Transaction/Methods.cpp:2350]: An alternative approach would be to explicitly check `if (tryRes.hasValue()) { /* handle response */ } else { /* log and fall through */ }`.

Comment (neunhoef) [reply arangod/Transaction/Methods.cpp:2350]: That is right. I do not actually know which I prefer. In any case, I think we had decided against the use of exceptions for error handling a long time ago...

**Comment by neunhoef** [comment]
We try a different approach which would get rid of the exception in the first place.

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11460/

**Comment by neunhoef** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11470/

File: arangod/Network/Methods.cpp [hunk 4]
```diff
@@ -111,6 +111,32 @@ auto prepareRequest(RestVerb type, std::string path, VPackBufferUInt8 payload,
...
 }
 
+/// @brief Function to produce a response object from thin air:
+static std::unique_ptr<fuerte::Response> buildResponse(fuerte::StatusCode statusCode, Result const& res) {
+  VPackBuffer<uint8_t> buffer;
+  VPackBuilder builder(buffer);
+  {
+    VPackObjectBuilder guard(&builder);
+    auto errorNum = res.errorNumber();
+    builder.add(StaticStrings::Error, VPackValue(errorNum != TRI_ERROR_NO_ERROR));
```
Comment (mchacki) [inline arangod/Network/Methods.cpp:121]: In the 3.7 back port of this PR there is comment about this Error generation here.

Please make sure it gets considered in both PRs.

**Comment by neunhoef** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11505/

File: arangod/Network/Methods.cpp (hunk 4 above)
Comment (neunhoef) [reply arangod/Network/Methods.cpp:121]: Did look at the comment, and did a change, here and over there.

**Pull request merged**